x-pack/legacy/plugins/siem/public/containers/detection_engine/rules/types.ts

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import * as t from 'io-ts';
+
+export const RuleSchema = t.type({
```

Comment:
I think that I will write my types like that to allow partial attributes in the rule object
```
export const RuleSchema = t.intersection([
  t.type({
    description: t.string,
    enabled: t.boolean,
    false_positives: t.array(t.string),
    from: t.string,
    index: t.array(t.string),
    interval: t.string,
    language: t.string,
    max_signals: t.number,
    name: t.string,
    query: t.string,
    references: t.array(t.string),
    severity: t.string,
    tags: t.array(t.string),
    to: t.string,
    type: t.string,
  }),
  t.partial({
    created_by: t.string,
    id: t.string,
    rule_id: t.string,
    updated_by: t.string,
  }),
]);
```

Reply:
I like that -- is this a fine partial for import validation though? Don't we need `to`/`from` as part of initial configuration as well?

Reply:
I put all the optional field in the `partial` but I can not promise that it won't change again

Reply:
I updated the schema as follows:

```ts
export const RuleSchema = t.intersection([
  t.type({
    created_by: t.string,
    description: t.string,
    enabled: t.boolean,
    id: t.string,
    index: t.array(t.string),
    interval: t.string,
    language: t.string,
    name: t.string,
    query: t.string,
    rule_id: t.string,
    severity: t.string,
    type: t.string,
    updated_by: t.string,
  }),
  t.partial({
    false_positives: t.array(t.string),
    from: t.string,
    max_signals: t.number,
    references: t.array(t.string),
    tags: t.array(t.string),
    to: t.string,
  }),
]);
```

We can have a separate `NewRuleSchema` that leaves off the server-generated fields like `created_at`, `id`, etc that you can be used for rule creation/importing. For resources returned from the rules API, we can use the above type. Though to be honest, the Rules API returns all fields filled in, so this partial probably isn't necessary if we also have the `NewRuleSchema`


Sample response:

``` ts
{
  "created_by": "elastic",
  "description": "Linux: Systemctl Activity",
  "enabled": true,
  "false_positives": [],
  "filters": [],
  "from": "now-6m",
  "id": "3db7e18b-979b-4dba-a321-2ebc00bfe133",
  "immutable": true,
  "index": [],
  "interval": "5m",
  "language": "kuery",
  "max_signals": 100,
  "name": "Linux: Systemctl Activity",
  "query": "process.name: systemctl",
  "references": [],
  "rule_id": "linux_systemctl_activity",
  "severity": "low",
  "tags": [],
  "to": "now",
  "type": "query",
  "updated_by": "elastic"
}

```